### PR TITLE
chore: Add warning about implicit logical ID usage in sam build

### PR DESCRIPTION
**Pull request by aahung**
#### Which issue(s) does this change fix?

https://github.com/aws/serverless-application-model/issues/2152

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Comment by hoffa** [comment]
Breadcrumb to related PR: https://github.com/aws/serverless-application-model/pull/2043

**Review by CoshUS — APPROVED** [review]

File: samcli/commands/build/command.py [hunk 1]
```diff
@@ -295,6 +296,8 @@ def do_cli(  # pylint: disable=too-many-locals, too-many-statements
         build_images=processed_build_images,
         aws_region=click_ctx.region,
     ) as ctx:
+        SamApiProvider.check_implicit_api_resource_ids(ctx.stacks)
```
Comment (hawflau) [inline samcli/commands/build/command.py:299]: Does it check whether the template is a SAM template? i.e. with `Transform: AWS::Serverless-2016-10-31`?

I believe SAM CLI also supports building pure CFN template. In that case, the implicit API logical IDs should not be reserved.

**Review by hawflau — CHANGES_REQUESTED** [review]

**Comment by aahung** [comment]
@CoshUS @hawflau I rebased this PR, moved the invocation from `do_cli()` to `BuildContext.run()`

File: samcli/commands/build/command.py (hunk 1 above)
Comment (aahung) [reply samcli/commands/build/command.py]: addressed

File: samcli/lib/providers/sam_api_provider.py [hunk 2]
```diff
@@ -327,3 +328,20 @@ def _get_route_stack_depth(route: Route) -> int:
...
+    def check_implicit_api_resource_ids(stacks: List[Stack]) -> None:
+        for stack in stacks:
+            for logical_id in stack.resources:
+                if logical_id in (
+                    SamApiProvider.IMPLICIT_API_RESOURCE_ID,
+                    SamApiProvider.IMPLICIT_HTTP_API_RESOURCE_ID,
+                ):
+                    LOG.warning(
+                        Colored().yellow(
+                            'Your template contains a resource with logical ID "%s," '
```
Comment (hoffa) [inline samcli/lib/providers/sam_api_provider.py:342]:
```suggestion
                            'Your template contains a resource with logical ID "%s", '
```

**Review by CoshUS — APPROVED** [review]

**Review by hawflau — APPROVED** [review]

**Review by hoffa — APPROVED** [review]

**Pull request merged**